ewatercycle/forcing.py

```diff
@@ -0,0 +1,341 @@
+from esmvalcore.experimental import get_recipe
+from pathlib import Path
+
+FORCINGS = {
+    'ERA5': {
+        'dataset': 'ERA5',
+        'project': 'OBS6',
+        'tier': 3,
+        'type': 'reanaly',
+        'version': 1
+    },
+    'ERA-Interim': {
+        'dataset': 'ERA-Interim',
+        'project': 'OBS6',
+        'tier': 3,
+        'type': 'reanaly',
+        'version': 1
+    },
+}
+
+
+def update_hype(
+    data: dict,
+    *,
+    forcings: list = None,
+    startyear: int = None,
+    endyear: int = None,
+    shapefile: str = None,
+):
+    """
+    Update hype recipe data in-place.
+    
+    Parameters
+    ----------
+    forcings : list
+        List of forcings to use
+    startyear : int
+        Start year for the observation data
+    endyear : int
+        End year for the observation data
+    shapefile : str
+        Name of the shapefile to use.
+    """
+    preproc_names = ('preprocessor', 'temperature', 'water')
+
+    if shapefile is not None:
+        for preproc_name in preproc_names:
+            data['preprocessors'][preproc_name]['extract_shape'][
+                'shapefile'] = shapefile
+
+    if forcings is not None:
+        datasets = [FORCINGS[forcing] for forcing in forcings]
+        data['datasets'] = datasets
+
+    variables = data['diagnostics']['hype']['variables']
+    var_names = 'tas', 'tasmin', 'tasmax', 'pr'
+
+    if startyear is not None:
+        for var_name in var_names:
+            variables[var_name]['start_year'] = startyear
+
+    if endyear is not None:
+        for var_name in var_names:
+            variables[var_name]['end_year'] = endyear
+
+    return data
+
+
+def update_lisflood(
+    data: dict,
+    *,
+    forcings: list = None,
+    startyear: int = None,
+    endyear: int = None,
+    shapefile: str = None,
+    extract_region: dict = None,
```

Comment:
In the case of listflood (and maybe other models as well), we use both extract region *and* extract shape. I think the extract region settings in these cases is based on the extent of the shapefile plus some padding. This is what we used to do in the preprocessing notebooks in the comparison. 

I think it would make sense to do that padding also inside the forcing module. That would make the interface for all models more uniform and save hassle for the hydrologist, who shouldn't have to worry about using multiple preprocessors to eventually get his shapefile cut out. 

For example, from the listflood preprocessing notebook in the comparison: 

```python
def get_extents(shapefile, pad=0):
    """Get lat/lon extents from shapefile and add padding."""
    shape = fiona.open(shapefile)
    x0, y0, x1, y1 = [
        shply.geometry.shape(p["geometry"]).bounds for p in shape
    ][0]
    x0 = round((x0 - pad), 1)
    y0 = round((y0 - pad), 1)
    x1 = round((x1 + pad), 1)
    y1 = round((y1 + pad), 1)
    return x0, y0, x1, y1
```

Reply:
I agree, but that is outside of scope for this PR. We will get back to this when we agree on the design for the API.